Lib/python/std_filesystem.i

```diff
@@ -0,0 +1,85 @@
+/*
+  filesystem::path
+*/
+%{
+#if __cplusplus >= 201703L
+#include <filesystem>
+#include <Python.h>
+#endif // __cplusplus >= 201703L
+%}
+
+%fragment("common", "header") {
+    PyObject * importPathCls() {
+        PyObject * module = PyImport_ImportModule("pathlib");
+        PyObject * cls = PyObject_GetAttrString(module, "Path");
+        Py_DECREF(module);
+        return cls;
+    }
+
+    bool isPathInstance(PyObject * obj) {
+        PyObject * cls = importPathCls();
+        bool is_instance =  PyObject_IsInstance(obj, cls);
+        Py_DECREF(cls);
+        return is_instance;
+    }
+
+    const char * pathToStr(PyObject * obj) {
+        PyObject * str_obj = PyObject_Str(obj);     // New reference
+        const char * s = PyUnicode_AsUTF8(str_obj); // This stores the UTF-8 representation buffer within str_obj
+        // Py_DECREF(str_obj);                      // So we don't decref here
+        return s;
+    }
+}
+
+%typemap(in, fragment="common") std::filesystem::path {
+    if (PyUnicode_Check($input)) {
+        const char* s = PyUnicode_AsUTF8($input);
+        $1 = std::filesystem::path(s);
+    }
+    else if (isPathInstance($input)) {
+        const char * s = pathToStr($input);
+        $1 = std::filesystem::path(s);
+    }
+    else {
+        void* argp = 0;
+        int res = SWIG_ConvertPtr($input, &argp, $descriptor(std::filesystem::path *), $disown |  0 );
+        if (!SWIG_IsOK(res)) {
+            %argument_fail(res, "$type", $symname, $argnum);
+        }
+        std::filesystem::path * temp = %reinterpret_cast(argp, $ltype*);
+        $1 = *temp;
+    }
+}
+
+%typemap(in, fragment="common") std::filesystem::path const *, std::filesystem::path const & {
+    if (PyUnicode_Check($input)) {
+        const char* s = PyUnicode_AsUTF8($input);
+        $1 = new std::filesystem::path(s);
+    }
+    else if (isPathInstance($input)) {
+        const char * s = pathToStr($input);
+        $1 = new std::filesystem::path(s);
+    }
+    else {
+        void* argp = 0;
+        int res = SWIG_ConvertPtr($input, &argp, $descriptor, $disown |  0 );
+        if (!SWIG_IsOK(res)) {
+            %argument_fail(res, "$type", $symname, $argnum);
+        }
+        $1 = %reinterpret_cast(argp, $ltype);
+    }
+}
+
+%typemap(out, fragment="common") std::filesystem::path {
+    const std::string& s = $1.string();
```

Comment:
I think it is actually here that we need to do something different.

Am I allowed to do an an `if constexpr` with type_traits here since this is C++17 only?

```
if constexpr (std::is_same_v< typename fs::path::value_type, wchar_t>) {
        // use generic_wstring() and PyUnicode_FromString
} else {
       // use generic_string() and PyUnicode_FromWideChar or somehing
}
```

The alternative is to use macros and assume WIN32 => wide string, otherwise string

Reply:
I've actually tested the rountripping of paths with special char on windows and it appears to work.

Reply:
The mingw WITHPYTHON=python is failing, all others windows CI builds are ok though... https://ci.appveyor.com/project/swig/swig/builds/47323349/job/8ru8l4v46ipumj8y#L1079

Reply:
going to try to catch it in d29d8947d962224be0b1e1f3fd4a64003909d6f2

Reply:
```
checking python testcase cpp17_director_string_view (with run test)
Traceback (most recent call last):
  File "C:\projects\swig\Examples\test-suite\python\cpp17_std_filesystem_runme.py", line 65, in <module>
    assert specialPath == roundTriped, "'{p}' != '{p2}'".format(p=specialPath, p2=roundTriped)
AssertionError: '/\u5bb6/\u5c4b' != '/家/屋'
make[1]: *** [Makefile:133: cpp17_std_filesystem.cpptest] Error 1
```

Reply:
> Am I allowed to do an an `if constexpr` with type_traits here since this is C++17 only?

That seems reasonable as these typemaps require a C++17 compiler for `std::filesystem::path` anyway.